src/mon/ConfigMonitor.cc

```diff
@@ -333,30 +334,37 @@ bool ConfigMonitor::preprocess_command(MonOpRequestRef op)
 	goto reply;
       }
       if (opt->has_flag(Option::FLAG_NO_MON_UPDATE)) {
-	// handle special options
-	if (name == "fsid") {
-	  odata.append(stringify(mon.monmap->get_fsid()));
-	  odata.append("\n");
-	  goto reply;
-	}
-	err = -EINVAL;
-	ss << name << " is special and cannot be stored by the mon";
-	goto reply;
+        // handle special options
+        if (name == "fsid") {
+          value = stringify(mon.monmap->get_fsid());
+        }
+        else {
+          err = -EINVAL;
+          ss << name << " is special and cannot be stored by the mon";
+          goto reply;
+        }
       }
       // get a single value
       auto p = config.find(name);
       if (p != config.end()) {
-	odata.append(p->second);
-	odata.append("\n");
-	goto reply;
+        value = p->second;
       }
-      if (!entity.is_client() &&
-	  opt->daemon_value != Option::value_t{}) {
-	odata.append(Option::to_str(opt->daemon_value));
+      else if (!entity.is_client() &&
+              opt->daemon_value != Option::value_t{} ) {
+        value = opt->daemon_value;
       } else {
-	odata.append(Option::to_str(opt->value));
+        value = opt->value;
+      }
+      if (!f) {
+        odata.append(Option::to_str(value));
+        odata.append("\n");
+      }
+      else {
+        f->open_object_section("option");
+        f->dump_string(name, Option::to_str(value));
+        f->close_section();
```

Comment:
```suggestion
        f->open_object_section("option");
        opt->dump_value("value", value, f);
        op->dump(f);
        f->close_section();
```

Reply:
Elsewhere we dump the `Option` so we should do so here too.

What we're missing here is "section" but I'm not sure we have that information.

Reply:
You can see in the history that I've originally used `opt->dump_value`, but then changed it. The reason is to conform with all other places around this code that dump options, which appear to report all values as string, see the outputs of dump/show config. There's also this awkward case of `null` being rendered as empty string because Formatter doesn't support null 🤷🏻 
Anyhow, I referred to this particular case in my comment: https://github.com/ceph/ceph/pull/52882#issuecomment-1671038579

Reply:
> You can see in the history that I've originally used `opt->dump_value`, but then changed it. The reason is to conform with all other places around this code that dump options, which appear to report all values as string, see the outputs of dump/show config.

That's unfortunate. I think we can add another field which dumps the value using `dump_value` but with another name. It improves usability of the json.

> There's also this awkward case of `null` being rendered as empty string because Formatter doesn't support null 🤷🏻 

That's no longer the case with https://github.com/ceph/ceph/pull/52547

> Anyhow, I referred to this particular case in my comment: [#52882 (comment)](https://github.com/ceph/ceph/pull/52882#issuecomment-1671038579)

Can you lay out every way an option is dumped as json in a comment so it's easier for us to consolidate in this discussion. I would like to make it consistent where possible while also giving the maximum amount of information.

Reply:
have you seen https://github.com/ceph/ceph/pull/52882#issuecomment-1670877239? That's where I tried to lay it all out

Reply:
>> There's also this awkward case of null being rendered as empty string because Formatter doesn't support null 🤷🏻
> That's no longer the case with https://github.com/ceph/ceph/pull/52547
This is great! However:
1. it's still not merged and has a bunch of other changes bundled
2. it's not backward compatible to update the existing dump code, and to add it only to the fixed flows will cause incoherence with the API which IMO is worse than an empty string for a null.

Reply:
CC @rzarzynski